include/Gaffer/Private/IECorePreview/TaskMutex.h

```diff
@@ -0,0 +1,356 @@
+/// spawned by the holder. Useful for performing expensive delayed
+/// initialisation of shared resources.
+///
+/// Based on an example posted by "Alex (Intel)" on the following thread :
+///
+/// https://software.intel.com/en-us/forums/intel-threading-building-blocks/topic/703652
+///
+/// Simple usage :
+///
+/// 	```
+/// 	void performExpensiveInitialisationUsingTBB();
+/// 	static bool g_initialised;
+/// 	static TaskMutex g_mutex;
+/// 	...
+/// 	TaskMutex::ScopedLock lock( g_mutex );
+/// 	if( !g_initialised )
+/// 	{
+/// 		lock.execute( []{ performExpensiveInitialisationUsingTBB(); } );
+/// 		g_initialised = true;
+/// 	}
+/// 	// Use resource here, while lock is still held.
+/// ```
+///
+/// Improved performance via reader locks :
+///
+/// 	```
+/// 	...
+/// 	// Optimistically take a reader lock, sufficient to allow us
+/// 	// to read from the resource if it is already initialised.
+/// 	TaskMutex::ScopedLock lock( g_mutex, /* write = */ false );
+/// 	if( !g_initialised )
+/// 	{
+/// 		// Upgrade to writer lock, so we can initialise the shared resource.
+/// 		lock.upgradeToWriter();
+/// 		if( !g_initialised ) // Check again, we may not be the first to get a write lock
+/// 		{
+/// 			lock.execute( []{ performExpensiveInitialisationUsingTBB(); } );
+/// 			g_initialised = true;
+///			}
+/// 	}
+/// 	// Use resource here, while lock is still held.
+/// ```
+class TaskMutex : boost::noncopyable
+{
+
+	typedef tbb::spin_rw_mutex InternalMutex;
+
+	public :
+
+		TaskMutex()
+		{
+		}
+
+		/// Used to acquire a lock on the mutex and release it
+		/// automatically in an exception-safe way. Equivalent to
+		/// the `scoped_lock` of the standard TBB mutexes.
+		class ScopedLock : boost::noncopyable
+		{
+
+			public :
+
+				ScopedLock()
+					:	m_mutex( nullptr ), m_writer( false ), m_recursive( false )
+				{
+				}
+
+				ScopedLock( TaskMutex &mutex, bool write = true, bool acceptWork = true )
+					:	ScopedLock()
+				{
+					acquire( mutex, write, acceptWork );
+				}
+
+				~ScopedLock()
+				{
+					if( m_mutex )
+					{
+						release();
+					}
+				}
+
+				/// Acquires a lock on `mutex`. If `acceptWork` is true, then may perform
+				/// work on behalf of `execute()` while waiting.
+				void acquire( TaskMutex &mutex, bool write = true, bool acceptWork = true )
+				{
+					tbb::internal::atomic_backoff backoff;
+					while( !acquireOr( mutex, write, [acceptWork]( bool workAvailable ){ return acceptWork; } ) )
+					{
+						backoff.pause();
+					}
+				}
+
+				/// Upgrades a previously-acquired reader lock to a full writer
+				/// lock.
+				bool upgradeToWriter()
+				{
+					assert( m_mutex && !m_writer && !m_recursive );
+					return m_lock.upgrade_to_writer();
+				}
+
+				/// Calls `f` in a way that allows threads waiting for the lock to perform
+				/// TBB tasks on its behalf. Should only be called by the holder of a write lock.
+				template<typename F>
+				void execute( F &&f )
```

Comment:
I definitely think that the header would be more readable if the implementation of the functions was squirrelled away in a `.inl` file, but I'd like to avoid that for the immediate future. I've made several changes while tidying this up for review, and want to retain the ability to diff it easily against the version that Murray has actually been testing in the wild.